Create "support" stream type with topic-specific permissions #19434

sanchi-t opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #19434 <!-- Issue link, or clear description.-->

This PR introduces support for creating private support-type streams, where topics are accessible only to specific user groups. Members of these designated groups can initiate new topics and participate in discussions within topics they create. Conversely, users outside these restricted groups can view all messages and topics within the stream.

To implement this functionality, a new stream setting named "who can access all topics?" is introduced. This setting is defined by a new stream property called `can_access_stream_topics_group`. By default, the `can_access_stream_topics_group` property will be assigned the value "role:everyone," indicating that all users will have access to all topics and messages within the stream.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sanchi-t committed:
message_send: Do not create `UserMessage` rows for restricted users.

The `check_access_based_on_can_access_stream_topics_group` function,
determines restricted users based on `is_user_in_group`.

In `do_send_messages`, `user_ids` are derived from `um_eligible_user_ids`
rather than `active_user_ids` for support streams. This ensures
that server events for new messages aren't sent to users who
lack access to those messages.

Using `can_access_stream_topics_group.named_user_group.name` in
`is_support_stream` function is causing the increase in database
query count.

Used `select_related` to fetch
`can_access_stream_topics_group.named_user_group`
to optimize database query count.

Also create UserMessage rows for idle users so that
no new `UserMessage` rows are made for soft-deactivated
users when they get reactivated.

sanchi-t committed:
can_access_stream_history: Deny stream history access.

We need to check that the user is not restricted in
order to allow them to view messages in support stream.

IncludeHistoryTest: Update test to include support streams.

get_topics_backend: Restrict topics for `support` streams.

sanchi-t committed:
stream: Add new fields to the stream and types.

`can_access_stream_topics_group` have been added
to various types to align with the stream model.

timabbott commented on web/src/stream_popover.js:
```diff
@@ -173,6 +173,7 @@ function build_stream_popover(opts) {
                 $(this).closest(".popover").fadeOut(500).delay(500).remove();
 
                 const sub = stream_popover_sub(e);
+                hide_stream_popover();
```
Is this an existing bug we should fix via a preparatory PR?

sanchi-t replied:
No no I accidentally added an unrelated commit while rebasing.

timabbott commented on zerver/migrations/0498_stream_stream_topic_view_policy.py:
```diff
@@ -0,0 +1,19 @@
+        migrations.AddField(
+            model_name="stream",
+            name="stream_topic_view_policy",
+            field=models.PositiveSmallIntegerField(default=STREAM_POST_POLICY_EVERYONE),
```
This should probably be a `UserGroup` foreign key.

sanchi-t replied:
I used `NamedUserGroup` instead. Check this out https://github.com/zulip/zulip/pull/29520#discussion_r1600628215.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -1448,6 +1454,14 @@ paths:
+                                    
+                                    Whether 
+
+                                    **Changes**: New in Zulip 5.0 (feature level 71).
```
How'd you get that version number? Be careful when copy-pasting :P

sanchi-t replied:
Yeah, I was just creating a template for myself to work on later. This PR wasn't ready for review back then.

timabbott commented on web/src/settings_config.ts:
```diff
@@ -1002,6 +1002,23 @@ export const stream_privacy_policy_values = {
+                "Must be invited by a subscriber; Topics created by others will be hidden from non-administrator users",
+        }),
+    },
+
```
This should likely be an independent setting, rather than a private type, in the UI.

I also generally recommend putting "add UI for creating a thing" commits at the end of a branch, after all the commits that make the feature work as designed.

sanchi-t replied:
Got it.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -430,30 +433,50 @@ def followed_topic_notification_recipients(setting: str) -> Set[int]:
     )
 
+    if stream and stream.is_support_stream and not message_sender.is_bot:
+        realm = get_realm_by_id(realm_id)
```
This is available as `stream.realm`; no need to fetch it separately.

timabbott commented on zerver/lib/topic.py:
```diff
@@ -321,3 +321,38 @@ def participants_for_topic(realm_id: int, recipient_id: int, topic_name: str) ->
+    
+    return first_message_in_topic.sender_id
+
+def check_access_based_on_stream_topic_view_policy(sender: UserProfile, stream: Stream) -> bool:
```
This should become a group membership check with the new groups-based permissions systems, which should be much cheaper and also a lot less semi-duplicated code to write.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -430,30 +433,50 @@ def followed_topic_notification_recipients(setting: str) -> Set[int]:
+            authorized_user_ids 
+            + ([topic_creator_user_id] if topic_creator_user_id is not None else []) 
+            + [sender_id]
+        )
```
This might be cleaner with just a `set.add(sender_id)` type line after initializing the starting set.

sanchi-t replied:
Done.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -1681,6 +1705,13 @@ def check_message(
+        if stream.is_support_stream and not check_access_based_on_stream_topic_view_policy(sender,stream):
+            topic_creator_user_id = get_topic_creator_user_id(realm.id,stream.recipient_id,topic_name)
+            if topic_creator_user_id and not (sender.id == topic_creator_user_id):
+                raise UnauthorizedTopicAccessError
```
So as written, you're calling `topic_creator_user_id` twice when sending a single message; you'll want to avoid that by just calling it once here, and passing it forward if `get_recipient_info` also needs the value.

sanchi-t replied:
Yeah fixed that.

timabbott commented:
I posted an initial review on this, but I think it probably makes sense for you to prioritize on other projects -- I'm not entirely sure after reading this that we have the right product design for this feature.

timabbott commented:
I talked to @alya a bit and I think we're closer to having the right product design for this than I'd thought, so feel free to work on this more -- the main reworking required is just using groups to avoid having to redo everything as part of #19525.

sanchi-t committed:
stream_create: Enable the creation of support streams.

sanchi-t commented on zerver/migrations/0521_stream_stream_topic_access_group.py:
```diff
@@ -0,0 +1,21 @@
+            model_name="stream",
+            name="stream_topic_access_group",
+            field=models.ForeignKey(
+                null=True, on_delete=django.db.models.deletion.RESTRICT, to="zerver.namedusergroup"
```
The only reason I am writing `NamedUserGroup` as foreign key instead of `UserGroup` is that I want to use to `is_support_stream` function defined [here](c7a8b3c045136acf0dc74b97979f5769632e2351).

With `UserGroup` there will be no `name` property on `stream_topic_access_group` and I will have to do something like this which I think will be a heavier database operation:

```
NamedUserGroup.objects.get(
  id=stream_topic_access_group_id, realm=realm, is_system_group=True
)
````

sanchi-t commented on zerver/models/streams.py:
```diff
@@ -187,6 +187,9 @@ def is_history_realm_public(self) -> bool:
     def is_history_public_to_subscribers(self) -> bool:
         return self.history_public_to_subscribers
 
+    def is_support_stream(self) -> bool:
+        return self.stream_topic_access_group.name != SystemGroups.EVERYONE
+
```
Accessing `stream_topic_access_group.name` is leading to an inflated number of database queries. Unfortunately, I can't find a way around.

sanchi-t commented on zerver/lib/user_groups.py:
```diff
@@ -464,12 +464,12 @@ def get_recursive_membership_groups(user_profile: UserProfile) -> QuerySet[UserG
 def is_user_in_group(
-    user_group: UserGroup, user: UserProfile, *, direct_member_only: bool = False
+    user_group: UserGroup, user_id: int, *, direct_member_only: bool = False
 ) -> bool:
     if direct_member_only:
-        return get_user_group_direct_members(user_group=user_group).filter(id=user.id).exists()
+        return get_user_group_direct_members(user_group=user_group).filter(id=user_id).exists()
 
-    return get_recursive_group_members(user_group=user_group).filter(id=user.id).exists()
+    return get_recursive_group_members(user_group=user_group).filter(id=user_id).exists()
```
The changes  were made to maintain the optimization outlined in function `get_recipient_info` in `message_send.py`, as shown below:

```
query: ValuesQuerySet[UserProfile, ActiveUserDict] = UserProfile.objects.filter(
            is_active=True
        ).values(
            "id",
            "enable_online_push_notifications",
            "enable_offline_email_notifications",
            "enable_offline_push_notifications",
            "is_bot",
            "bot_type",
            "long_term_idle",
        )
```
As later in the function `get_recipient_info` we need to call `check_access_based_on_stream_topic_access_group` which would then call `is_user_in_group`.

sanchi-t commented on zerver/actions/message_send.py:
```diff
@@ -393,14 +404,6 @@ def followed_topic_notification_recipients(setting: str) -> Set[int]:
-            "is_bot",
-            "bot_type",
-            "long_term_idle",
         )
```
To use this optimization I had to make changes in `is_user_in_group` function. Check this comment out https://github.com/zulip/zulip/pull/29520#discussion_r1600646775.

sanchi-t replied:
This is not relevant I suppose.

sanchi-t commented on zerver/tests/test_subs.py:
```diff
@@ -612,6 +617,93 @@ def test_can_remove_subscribers_group_on_stream_creation(self) -> None:
             "'can_remove_subscribers_group' setting cannot be set to 'role:nobody' group.",
         )
 
+    def test_stream_topic_access_on_stream_creation(self) -> None:
```
This test is almost same as ` test_can_remove_subscribers_group_on_stream_creation`. Is that okay?

sanchi-t replied:
Yeah I can do that in follow up.

sanchi-t commented on zerver/lib/user_groups.py:
```diff
@@ -283,12 +283,15 @@ def access_user_group_for_setting(
     permission_configuration: GroupPermissionSetting,
     current_setting_value: Optional[UserGroup] = None,
-) -> UserGroup:
+    is_named_user_group: Optional[bool] = False,
+) -> Union[UserGroup, NamedUserGroup]:
     if isinstance(setting_user_group, int):
         named_user_group = access_user_group_by_id(setting_user_group, user_profile, for_read=True)
         check_setting_configuration_for_system_groups(
             named_user_group, setting_name, permission_configuration
         )
+        if is_named_user_group:
+            return named_user_group
```
Not sure if this is the best approach.

sanchi-t committed:
streams: Add new field to stream model.

The `can_access_stream_topics_group` field have been
added to the stream model and across various types.

The new fields causes the query length to be increased
in `test_subs`

migrations: Add `can_access_stream_topics_group` field to `stream` model.
migrations: Set default value for `can_access_stream_topics_group`.
migrations: Alter `can_access_stream_topics_group` field.
Alter the `can_access_stream_topics_group` field to disallow `null` values.

do_import_realm: Update function to support
`can_access_stream_topics_group`.

sanchi-t commented on zerver/models/streams.py:
```diff
@@ -186,6 +186,9 @@ def is_history_realm_public(self) -> bool:
     def is_history_public_to_subscribers(self) -> bool:
         return self.history_public_to_subscribers
 
+    def is_support_stream(self) -> bool:
+        return self.stream_topic_access_group.named_user_group.name != SystemGroups.EVERYONE
+
```
Accessing `stream_topic_access_group.named_user_group.name` is leading to an inflated number of database queries. Unfortunately, I can't find a way around.

Making `stream_topic_access_group` be foreign key on `NamedUserGroup` will reduce the db query by 1.

sahil839 replied:
You can use `select_related` to fetch these fields in the query which fetches the stream.

sanchi-t replied:
Done.

sahil839 commented on zerver/migrations/0539_alter_stream_stream_topic_access_group.py:
```diff
@@ -0,0 +1,22 @@
+                related_name="+",
+                to="zerver.usergroup",
+            ),
+        ),
```
The migration files should be added with the commit that add the fields in model and does other changes to make sure individual commit pass tests.

sahil839 commented on zerver/lib/import_realm.py:
```diff
@@ -225,6 +225,15 @@ def fix_streams_can_remove_subscribers_group_column(data: TableData, realm: Real
+    )
+    for stream in data[table]:
+        stream["stream_topic_access_group"] = everyone_group
+
```
These changes should also be in the previous commit to make sure each commit pass tests individually.

sanchi-t replied:
Fixed

sahil839 commented on zerver/models/streams.py:
```diff
@@ -186,6 +186,9 @@ def is_history_realm_public(self) -> bool:
     def is_history_public_to_subscribers(self) -> bool:
         return self.history_public_to_subscribers
 
+    def is_support_stream(self) -> bool:
```
I guess this commit will fail coverage because this fucntion is not used anywhere as of this function.

sanchi-t replied:
Fixed.

sahil839 commented on zerver/tests/test_users.py:
```diff
@@ -908,7 +908,7 @@ def test_create_user_with_multiple_streams(self) -> None:
         prereg_user = PreregistrationUser.objects.get(email="fred@zulip.com")
 
-        with self.assert_database_query_count(80):
+        with self.assert_database_query_count(90):
```
We should try if we can decrease the queries here.

sanchi-t replied:
The issue with creating a `PreregistrationUser` is that we cannot pass streams during its creation. Instead, we must add streams later using `prereg_user.streams.set()`. However, this method does not actually set streams with fetched `select_related` fields.
```
  prereg_user = PreregistrationUser(
  email=email,
  referred_by=user_profile,
  invited_as=invite_as,
  realm=realm,
  include_realm_default_subscriptions=include_realm_default_subscriptions,
  notify_referrer_on_join=notify_referrer_on_join,
  )
  prereg_user.save()
  stream_ids = [stream.id for stream in streams]
  prereg_user.streams.set(stream_ids)
```

sahil839 reviewed (commented):
I have not reviewed the complete PR, but have posted some comments.

I think the commits are not in a state where each commit would pass tests, so you would need to fix that. Have posted a couple of comments about what to fix, but there might be more such things, so would be good if you check the complete PR yourself once.

sahil839 commented:
Also, I am not sure whether we want to allow to set the setting to multiple groups and non-system groups. Would be good to discuss it first and then update the PR accordingly rather than migrating the setting later.

sanchi-t committed:
echo: Avoid locally echoing a message.

We need to check if a restricted user can access the
topic before echoing their message, which happens on
the server side.

sanchi-t commented on zerver/tests/test_message_fetch.py:
```diff
@@ -1157,6 +1160,28 @@ def test_ok_to_include_history(self) -> None:
+        self.assertFalse(ok_to_include_history(narrow, user_profile, False))
+        self.assertFalse(ok_to_include_history(narrow, guest_user_profile, False))
+        self.assertFalse(ok_to_include_history(narrow, subscribed_user_profile, False))
+        self.assertTrue(ok_to_include_history(narrow, admin_user_profile, False))
```
Done.

sanchi-t commented on zerver/openapi/zulip.yaml:
```diff
@@ -22515,6 +22554,18 @@ components:
+        system groups, except for the system groups named
+        `"role:internet"`, `"role:owners"` and `"role:nobody"`.
+
+        **Changes**: New in Zulip 9.0 (feature level 262).
```
Corrected.

sanchi-t commented on zerver/models/streams.py:
```diff
@@ -135,6 +135,11 @@ class Stream(models.Model):
     # We are not using PROTECT since we want to allow deletion of user groups
     # when realm itself is deleted.
     can_remove_subscribers_group = models.ForeignKey(UserGroup, on_delete=models.RESTRICT)
+    stream_topic_access_group = models.ForeignKey(
```
Done.

sanchi-t commented on zerver/migrations/0538_set_default_for_stream_topic_access_group.py:
```diff
@@ -0,0 +1,33 @@
+        )
+        Stream.objects.filter(realm=realm, stream_topic_access_group=None).update(
+            stream_topic_access_group=everyone_group
+        )
```
Done.

sanchi-t commented on zerver/tests/test_subs.py:
```diff
@@ -5585,15 +5771,15 @@ def test_subscriptions_query_count(self) -> None:
 
         # Test creating a public stream when realm does not have a notification stream.
-        with self.assert_database_query_count(38):
+        with self.assert_database_query_count(40):
```
This creates an extra query when creating a new stream using the following code snippet:
```
(stream, created) = Stream.objects.get_or_create(
        realm=realm,
        name__iexact=stream_name,
        defaults=dict(
            name=stream_name,
            creator=acting_user,
            description=stream_description,
            invite_only=invite_only,
            is_web_public=is_web_public,
            stream_post_policy=stream_post_policy,
            history_public_to_subscribers=history_public_to_subscribers,
            is_in_zephyr_realm=realm.is_zephyr_mirror_realm,
            message_retention_days=message_retention_days,
            can_remove_subscribers_group=can_remove_subscribers_group,
        ),
    )
```

How can I modify it so that when creating a new stream, it returns the stream with `can_remove_subscribers_group.namedusergroup.name`?

sanchi-t replied:
No it did not helped.

sahil839 reviewed (commented):
Posted some comments. This still some improvements in the commit structure to make sure that tests pass for each commit.

Also, we can reduce some code duplication - like setting the default of settings while creating stream and for the dropdown widgets JS code in stream creation codepath. But that can be done as follow-up.

sahil839 commented:
Also, a couple of questions -
- What happens if a user who is not part of the "Who can access all topics" group is mentioned in a new topic in a support stream? Does this user gains access to all messages of a topic or just that message?
- What happens when a soft-deactivated user becomes active and the `UserMessage` rows are created for them? Does that code take into account the stream being a support stream?

sanchi-t commented on zerver/views/streams.py:
```diff
@@ -244,6 +244,9 @@ def update_stream_backend(
     ),
+    can_access_stream_topics_group_id: int | None = REQ(
+        "can_access_stream_topics_group", json_validator=check_int, default=None
+    ),
```
Done.

sanchi-t commented on zerver/tests/test_subs.py:
```diff
@@ -4934,7 +4934,7 @@ def test_multi_user_subscription(self) -> None:
         with (
             self.capture_send_event_calls(expected_num_events=5) as events,
-            self.assert_database_query_count(38),
+            self.assert_database_query_count(40),
```
Done.

sanchi-t commented on web/tests/lib/events.js:
```diff
@@ -54,6 +54,7 @@ exports.test_streams = {
         message_retention_days: null,
         stream_post_policy: 1,
         can_remove_subscribers_group: 2,
+        can_access_stream_topics_group: 14,
```
Done.

sanchi-t commented on web/templates/stream_settings/stream_types.hbs:
```diff
@@ -71,6 +71,11 @@
+        {{> ../dropdown_widget_with_label
+          widget_name=stream_topic_access_setting_widget_name
+          label=(t 'Who can see all the topics?')
+          value_type="number"}}
```
Was following advice from [this](https://github.com/zulip/zulip/pull/29520#discussion_r1573388162) comment. But I can make changes if you would like me to.

sahil839 replied:
Yes, but it is fine to include the JS code for UI in the commit which adds the template.

sanchi-t replied:
Done.

sanchi-t commented on web/src/stream_settings_ui.js:
```diff
@@ -186,6 +186,12 @@ export function update_can_remove_subscribers_group_id(sub, new_value) {
+export function update_can_access_stream_topics_group_id(sub, new_value) {
+    stream_data.update_can_access_stream_topics_group_id(sub, new_value);
+    stream_ui_updates.update_setting_element(sub, "can_access_stream_topics_group");
+    stream_edit_subscribers.rerender_subscribers_list(sub);
```
Yeah there was no need. Removed.

sanchi-t commented on zerver/lib/message.py:
```diff
@@ -397,7 +397,11 @@ def is_subscribed_helper() -> bool:
+    if (
+        stream.is_public()
+        and not stream.is_support_stream()
+        and user_profile.can_access_public_streams()
```
No you are right. Fixed the logic.

sanchi-t commented:
> What happens if a user who is not part of the "Who can access all topics" group is mentioned in a new topic in a support stream? Does this user gains access to all messages of a topic or just that message?

The user does not gain access to that message or topic. The user neither get any notification.

> What happens when a soft-deactivated user becomes active and the UserMessage rows are created for them? Does that code take into account the stream being a support stream?

No the code did not take that into account. I now have included logic to handle that.

> For the groups to be allowed in the settings, if we want to allow all types of groups in future, we can probably just allow them in dev and tests for now and only show system groups in UI as we do for other settings.

Based on this message:
"I think for all groups settings, it is simplest to have new settings that we're merging only allow system groups aka the things equivalent to roles."

in the [design discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/Support.20Streams/near/1851932), I don't think we want to allow all types of groups in future.

sanchi-t commented on zerver/lib/soft_deactivation.py:
```diff
@@ -201,6 +204,11 @@ def add_missing_messages(user_profile: UserProfile) -> None:
                 # We are going to short circuit this iteration as its no use
                 # iterating since user unsubscribed before soft-deactivation
                 continue
+
+        if stream.is_support_stream() and not check_access_based_on_can_access_stream_topics_group(
+            user_profile, stream
+        ):
+            continue
```
I don't know if this is the right approach but I couldn't think of anything else. One thing that came to my mind was doing something like this:
```
subscription_logs = list(
        RealmAuditLog.objects.filter(
            modified_user=user_profile, 
            modified_stream_id__in=stream_ids, 
            modified_stream__can_access_stream_topics_group_id=everyone_group_id,
            event_type__in=events
        )
        .order_by("event_last_message_id", "id")
        .only("id", "event_type", "modified_stream_id", "event_last_message_id")
    )
```

But this is wrong because this won't create `UserMessage` rows for a soft deactivated user who is in `can_access_stream_topics_group` in a support stream.

sanchi-t replied:
This is also increasing the query count.

sanchi-t replied:
I just noticed that this doesn't account for the scenario where a user who is not a member of `can_access_stream_topics_group` receives messages in a topic they created while they are soft-deactivated. Ideally, we should create `UserMessage` rows for all messages sent in that topic. However, this would require fetching the first message in the topic to identify the sender, which could result in additional queries. Should we proceed with this approach despite the extra queries?

sanchi-t replied:
Right I will ask CZO.

sanchi-t replied:
Based on the [CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Support.20Streams/near/1918573) discussion the soft-reactivation should skip support streams.

sanchi-t committed:
filter_by_subscription_history: Remove unused parameter.

sahil839 reviewed (commented):
Posted some comments.

> Based on this message:
"I think for all groups settings, it is simplest to have new settings that we're merging only allow system groups aka the things equivalent to roles."
> in the [design discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/Support.20Streams/near/1851932), I don't think we want to allow all types of groups in future.

I understood this as we would allow system groups for now but we can add backend support for other groups for future. I see no reason to not allow other groups in future.

sanchi-t commented on zerver/tests/test_invite.py:
```diff
@@ -146,7 +146,7 @@ def test_query_counts_when_admin_assigns_streams(self) -> None:
         new_user = self.create_simple_new_user(realm, new_user_email)
 
-        with self.assert_database_query_count(14):
+        with self.assert_database_query_count(22):
```
Yeah because we can't use `select_related` for streams in `PreregistrationUser` to fetch `can_access_stream_topics_group__named_user_group` but I tried `prefetch_related` and it worked. Is it okay to use?

sahil839 replied:
Yes, that's because streams is a many to many field and it is fine to use `prefetch_related`.

sanchi-t commented on zerver/actions/message_send.py:
```diff
@@ -440,6 +450,29 @@ def get_ids_for(f: Callable[[ActiveUserDict], bool]) -> set[int]:
         lambda r: not r["is_bot"] or r["bot_type"] not in UserProfile.SERVICE_BOT_TYPES,
     )
 
+    if stream and is_support_stream and not message_sender.is_bot:
```
Yeah so the idea was that I thought messages by bot should be visible to everyone incase we want to communicate with all the members of a support stream. Moreover if a bot tries to send a message in an unauthorized topic (checked by seeing if `bot_owner` have access) we throw `UnauthorizedTopicAccessError`. But this code would make 'UserMessage' rows for everyone which is wrong, I fixed it.

I should ask in CZO for how bots should behave.

sahil839 replied:
What we usually do is to check the permission for bot based on its owner. So, that might be the correct way, but would be good to confirm it once.

sanchi-t commented on zerver/lib/message.py:
```diff
@@ -397,6 +402,11 @@ def is_subscribed_helper() -> bool:
         ).exists()
 
+    if stream.is_support_stream() and not check_access_based_on_can_access_stream_topics_group(
+        user_profile, stream
```
Right, updated.

sanchi-t replied:
But it increases query count.

sanchi-t commented on zerver/lib/message.py:
```diff
@@ -474,6 +484,11 @@ def bulk_access_stream_messages_query(
+    if stream.is_support_stream() and not check_access_based_on_can_access_stream_topics_group(
+        user_profile, stream
+    ):
+        return Message.objects.none()
```
Fixed.

sahil839 commented:
A couple of more points -
- While testing I saw that a user who is not part of the `can_access_stream_topics_group` can see all the topics in left sidebar. I don't think we want that.
- What happens, if a user was initially part of the `can_access_stream_topics_group` group and send a message to a topic, but not started the topic, and then that user is not part of `can_access_stream_topics_group`. Can the user see the topic, i.e. atleast the message that they sent themselves? Or also could the message for which they have `UserMessage` rows, which they would have for the messages that were sent when user part of the group.

alya commented:
> While testing I saw that a user who is not part of the can_access_stream_topics_group can see all the topics in left sidebar. I don't think we want that.

Agreed.

alya commented:
> What happens, if a user was initially part of the `can_access_stream_topics_group` group and send a message to a topic, but not started the topic, and then that user is not part of `can_access_stream_topics_group`. Can the user see the topic, i.e. atleast the message that they sent themselves? Or also could the message for which they have `UserMessage` rows, which they would have for the messages that were sent when user part of the group.

I think the user should have access to messages based on their current permissions, regardless of their prior permissions. This is the same as for a user who is no longer subscribed to a private stream, and loses access to the messages they sent and received there.

sanchi-t committed:
create_test_message: Refactor function to accept value from a parameter.

This change will be useful for adding support stream scenario
in the later commit.
Additionally, redundant function calls for "topic1" have been
removed.

sahil839 commented on zerver/lib/message.py:
```diff
@@ -229,6 +236,19 @@ def messages_for_ids(
+                stream.is_support_stream()
+                and not check_access_based_on_can_access_stream_topics_group(user_profile, stream)
+            ):
+                topic_creator_user_id = get_topic_creator_user_id(
```
Similarly, doing these queries in a loop is also a problem. Is there a way where we can just filter out the inaccessible messages initially when fetching message IDs?

sanchi-t replied:
Here are [the](https://chat.zulip.org/#narrow/stream/101-design/topic/Support.20Streams/near/1943912) problems I faced when filtering out the inaccessible messages initially when fetching message IDs

sahil839 commented on zerver/lib/soft_deactivation.py:
```diff
@@ -191,6 +192,7 @@ def add_missing_messages(user_profile: UserProfile) -> None:
     recipient_ids = []
     for sub in all_stream_subs:
         stream_subscription_logs = all_stream_subscription_logs[sub["recipient__type_id"]]
+        stream = get_stream_by_id_in_realm(sub["recipient__type_id"], user_profile.realm)
```
I guess this is incorrect as of now, since the `UserMessage` rows are not created for soft-deactivated users when a message is sent, which needs to be handled in `get_recipient_info` and then this code will not be needed since the `UserMessage` already exists and those messages will be skipped.

sahil839 commented on zerver/views/streams.py:
```diff
@@ -560,6 +560,9 @@ def add_subscriptions_backend(
         Json[int | None], ApiParamConfig("can_remove_subscribers_group")
     ] = None,
+    can_access_stream_topics_group_id: Annotated[
+        Json[int | None], ApiParamConfig("can_access_stream_topics_group")
```
I think we had decided to allow the anonymous groups as well here. You can see `can_mention_group` setting used for a user group as a reference.

sanchi-t replied:
Yes, I updated it. I have a question regarding this: If can_access_stream_topics_group_id is an anonymous group, how will we determine if the stream is a support stream? Currently, we check it with the following condition:
```
can_access_stream_topics_group.named_user_group.name != SystemGroups.EVERYONE

```

sahil839 reviewed (commented):
Posted some comments.

sanchi-t commented on zerver/tests/test_invite.py:
```diff
@@ -142,11 +143,18 @@ def test_query_counts_when_admin_assigns_streams(self) -> None:
+        )
+
+        # Prefetch related Stream objects for the PreregistrationUser
+        prereg_user = PreregistrationUser.objects.prefetch_related(
```
`get_object_from_key` is responsible for getting the `PreregistrationUser` object. However I think this is all can be done in the function:
```
try:
        confirmation = Confirmation.objects.select_related("content_object").get(
            confirmation_key=confirmation_key, type__in=confirmation_types
        )
```

Not sure how I will prefetch streams with `can_access_stream_topics_group__named_user_group` for `PreregistrationUser`.

sahil839 replied:
So, we shouldn't prefetch it while testing as well to test the actual query count.

sahil839 replied:
We can do a separate query to fetch all the streams with required fields using `select_related` by using ID from the `PreregistrationUser` object.

sanchi-t replied:
Done.

sanchi-t committed:
test_user_groups: Unarchive stream for next iteration.

Unarchiving the stream ensures that the next iteration of the loop
receives an active stream, leading to the expected results.

sanchi-t commented on zerver/lib/message.py:
```diff
@@ -229,6 +236,19 @@ def messages_for_ids(
         msg_dict = message_dicts[message_id]
+        if user_profile and msg_dict["recipient_type"] == Recipient.STREAM:
+            stream_id = msg_dict["recipient_type_id"]
+            stream = get_stream_by_id_in_realm(stream_id, realm)
```
Yes here is [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/Support.20Streams/near/1943912) discussion regarding this.
